runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -5870,128 +5906,124 @@ TR::Register *J9::Power::TreeEvaluator::VMmonexitEvaluator(TR::Node *node, TR::C
-      generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, ctempReg, 0);
-      generateTrg1Src2Instruction(cg,TR::InstOpCode::Op_cmp, node, condReg, monitorReg, metaReg);
-
-      if (cg->comp()->target().cpu.id() >= TR_PPCgp)
-         // use PPC AS branch hint
-         generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, PPCOpProp_BranchUnlikely, node, decLabel, condReg);
-      else
-         generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, decLabel, condReg);
+      generateTrg1MemInstruction(cg, loadOpCode, node, monitorReg, new (cg->trHeapMemory()) TR::MemoryReference(baseReg, lwOffset, lockSize, cg));
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 0);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::Op_cmpl, node, condReg, monitorReg, metaReg);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, decrementCheckLabel, condReg);
 
       // exiting from read monitors still needs lwsync
       // (ensures loads have completed before releasing lock)
       if (cg->comp()->target().isSMP())
          generateInstruction(cg, TR::InstOpCode::lwsync, node);
-      if (cg->comp()->target().is64Bit() && !fej9->generateCompressedLockWord())
-         opCode = TR::InstOpCode::std;
-      else
-         opCode = TR::InstOpCode::stw;
-      generateMemSrc1Instruction(cg, opCode, node, new (cg->trHeapMemory()) TR::MemoryReference(baseReg, moffset, lockSize, cg), ctempReg);
+
+      generateMemSrc1Instruction(cg, storeOpCode, node, new (cg->trHeapMemory()) TR::MemoryReference(baseReg, lwOffset, lockSize, cg), tempReg);
       generateLabelInstruction(cg, TR::InstOpCode::b, node, doneLabel);
 
-      generateLabelInstruction(cg, TR::InstOpCode::label, node, decLabel);
+      generateLabelInstruction(cg, TR::InstOpCode::label, node, decrementCheckLabel);
       if (cg->comp()->target().is64Bit())
          {
-         generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, threadReg, OBJECT_HEADER_LOCK_RECURSION_MASK);
-         generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, threadReg, monitorReg, threadReg);
+         generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, tempReg, monitorReg, tempReg);
          }
       else
-         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, threadReg, monitorReg, 0, ~OBJECT_HEADER_LOCK_RECURSION_MASK);
-      generateTrg1Src2Instruction(cg, TR::InstOpCode::Op_cmpl, node, condReg, threadReg, metaReg);
+         {
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, tempReg, monitorReg, 0, ~LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK);
+         }
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::Op_cmpl, node, condReg, tempReg, metaReg);
```

Comment:
Given the pending integration for a unified driver from @gacholio ,  the choice of certain opCode(s) had better to be compile-time dynamic, instead of Op_cmpl way of being statically decided by #if.  I would strongly suggest you go over the whole code you modified at least to catch all the cases.

Reply:
I forgot to mention earlier but I've changed this. Instead of using variations of `Op_cmp`, there is a runtime check for compressed lockwords and whether it is 32bit or 64bit to select the cmp instruction to use. I only made the changes to the code sections I modified so there are still some uses of `Op_cmp` in lockword related code paths (eg. old reservation path) that I didn't update.